File: 1834.single-threaded-cpu/solution.kt
import java.util.PriorityQueue
import java.util.Comparator
data class Task(
    val enqueue: Int,
    val processTime: Int,
    val idx: Int
)
class Solution {
    fun getOrder(tasks: Array<IntArray>): IntArray {

        val results = ArrayList<Int>()
        val taskList = ArrayList<Task>()
        tasks.forEachIndexed { idx, task -> taskList.add(Task(
            task[0],
            task[1],
            idx
        ))}

        // sort by enqueue time
        taskList.sortWith( compareBy { it.enqueue })

        var i = 0
        var taskFinishTime = 0
        var heapQ = PriorityQueue<Task>( compareBy { it: Task -> it.processTime }.thenBy { it: Task -> it.idx })
        val N = taskList.size

        while ((i < N) || (!heapQ.isEmpty())) {
            if (heapQ.isEmpty() &&(taskFinishTime <= taskList.get(i).enqueue)) {
                taskFinishTime = taskList.get(i).enqueue
            }

            while ((i < N) && (taskList.get(i).enqueue <= taskFinishTime)) {
                heapQ.add(taskList.get(i))
                i += 1
            }

            val poll = heapQ.poll()
            taskFinishTime += poll.processTime
            results.add(poll.idx)
        }

        return results.toIntArray()
    }
}

File: 1834.single-threaded-cpu/solution.py
class Solution:
    def getOrder(self, tasks: List[List[int]]) -> List[int]:
        results = []
        N = len(tasks)

        # adding each indicies, then sort by enqueue times
        taskList = [(enqeue, process_time, idx) for idx, (enqeue, process_time) in enumerate(tasks)]
        taskList.sort()

        # min heap queue
        hq = []
        curr_time = 0
        i = 0
        while (i < N) or hq:

            # If all tasks less are done before next task,
            # move on to the next task
            if (not hq) and (curr_time < taskList[i][0]):
                curr_time = taskList[i][0]

            # Pick up all tasks less than the current task finish time
            while (i < N) and (taskList[i][0] <= curr_time):
                heapq.heappush(hq, (taskList[i][1], taskList[i][2]))
                i += 1

            # Extends the current task finish time
            poll = heapq.heappop(hq)
            curr_time += poll[0]
            results.append(poll[1])

        return results
